### PR TITLE
Writing Token Contract with F*

### PR DESCRIPTION
**Pull request by mariari**
Currently this contract is not complete, but it should have enough form to understand the approach

### CODE REVIEW

File: experimental/examples/token/Token.fst [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+let add_account_values accounts =
+ MapProp.fold (fun _key (v : nat) (acc : nat) -> v + acc) accounts 0
+
+
+unopteq type storage = {
+  total_supply : nat;
+  accounts : a : accounts { total_supply = add_account_values a };
+}
+
+val emtpy_storage : storage
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]:
```suggestion
val empty_storage : storage
```

File: experimental/examples/token/Token.fst [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+
+val has_n : accounts -> address -> nat -> bool
+let has_n accounts add to_transfer =
+  match Map.select add accounts with
+  | Some n -> to_transfer <= n
+  | None   -> false
+
+// useless lemma, thought i needed it, but I just did to_transfer >= n.
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]: Shall we delete this then?

File: experimental/examples/token/Token.fst [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+val transfer_add : acc : accounts
+                 -> add : address
+                 -> num : nat
+                 -> Lemma
+                  (ensures ( add_account_values acc + num
+                           == add_account_values (account_add acc add num)))
+
+let transfer_add acc add num =
+  match Map.select add acc with
+  | Some balance -> admit ()
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]: Looks like this needs a proof?

File: experimental/examples/token/Token.fst [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                 -> num : nat
+                 -> Lemma
+                  (requires (has_n acc add num))
+                  (ensures (  add_account_values acc - num
+                           == add_account_values (account_sub acc add num)))
+let transfer_sub acc add num =
+  match Map.select add acc with
+  | Some balance ->
+    let remaining : nat = balance - num in
+    admit ()
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]: Looks like this needs a proof?

**Comment by cwgoes** [comment]
Let's also add another two transaction types, `mint` and `burn`, which *do* change the total supply, but only when authorised by a particular owner (another field in the token contract storage).

File: experimental/examples/token/Token.fst (hunk 2 above)
Comment (mariari) [reply experimental/examples/token/Token.fst]: Yeah, will this commit

File: experimental/examples/token/Token.fst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        accounts     = account_sub token.storage.accounts burn_from_account burn_amount
+      }}
+
+type transaction_error =
+  | Not_enough_funds
+  | Not_same_account
+  | Not_owner_token
+  | Not_enough_tokens
+
+val execute_transaction : token -> tx -> c_or transaction_error token
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]: We should prove properties of this top-level function

File: experimental/examples/token/Token.fst [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+let execute_transaction token tx =
+  match tx.tx_data with
+  | Transfer _ ->
+    if valid_transfer token tx
```
Comment (cwgoes) [inline experimental/examples/token/Token.fst]: This reads storage once more than necessary

Comment (cwgoes) [reply experimental/examples/token/Token.fst]: or perhaps leave this out of F* for now but take notes on the general problem

File: experimental/examples/token/Token.fst (hunk 5 above)
Comment (mariari) [reply experimental/examples/token/Token.fst]: That is done with the newest push

File: experimental/examples/token/Token.fst (hunk 6 above)
Comment (mariari) [reply experimental/examples/token/Token.fst]: We should probbaly leave it out, as it's not too exciting without say a reader monad (F* has them I think, just lacking effect polymorphism is a pain!)

A future PR will be had adding it to the documents.

Which folder should we use?

Comment (cwgoes) [reply experimental/examples/token/Token.fst]: I think this is relevant to a standard library or examples section of language design choices - https://github.com/cryptiumlabs/juvix/pull/312.

**Pull request merged**